### ISSUE TITLE
Proposal: SemiPartial

### ISSUE THREAD

**Issue by DomParfitt** [krzkaczor/ts-essentials#77]
Would there be any interest in adding the below type?

```typescript
type SemiPartial<T, K extends keyof T> = Partial<Exclude<T, K>> & Pick<T, K>;
```

The idea behind it is to allow for objects where all the fields are optional (i.e. Typescript's `Partial<T>`) except for those that are specified.

For example:
```typescript
interface Foo {
    foo: string;
    bar: string;
    baz: string;
}

let foo: SemiPartial<Foo, 'foo' | 'bar'>;

// VALID - 'foo' & 'bar' are non-optional, 'baz' is optional
foo = {
    foo: "foo",
    bar: "bar",
};

// INVALID - both 'foo' and 'bar' must be present
foo = {
    baz: "baz",
};

// INVALID - 'bar' is present but 'foo' is not
foo = {
    bar: "bar",
    baz: "baz",
};
```

The main use I had in mind for this is working with GraphQL types. Typically properties on GraphQL types are optional by default and can be marked as required (using the `!` operator). In this respect it's the opposite of Typescript. This allows for a subset of that type to be provided when calling queries/mutations. 

For example, in the below a User is identified using an `email` (which is marked as required) but the other properties are optional for the purpose of updating. This means that, as an example, the `firstName` can be updated by just providing the `email` and `firstName` and the `lastName` is not required.

```graphql
type User {
  email: String!
  firstName: String
  lastName: String
}

mutation {
  updateUser(user: User)
}
```

You could define your Typescript interfaces to exactly match the GraphQL schema but there's a strong possibility that the domain model has different optionality requirements. For example, in the above it may be that both `firstName` and `lastName` are required fields for a `User` but optional for GraphQL queries/mutations. So your code might have something like:

```typescript
// This would be the 'true' domain model
interface User {
  email: string;
  firstName: string;
  lastName: string;
}

// This would be the type that is used to pass details to GraphQL queries/mutations
type GraphQLUser = SemiPartial<User, 'email'>
```

I've had  a look at the `MarkRequired` type that was discussed and implemented [here](https://github.com/krzkaczor/ts-essentials/issues/45), and this differs in that it makes non-optional properties optional unless specified, whereas `MarkRequired` only affects properties that were already optional.

Let me know what you think!

**Comment by quezak** [comment]
> `let foo: SemiPartial<Foo, 'foo' | 'bar'>;`

I think you can do the same with a combination of `MarkRequired` and `Partial`. Does this satisfy your use case?

```typescript
let foo: MarkRequired<Partial<Foo>, 'foo' | 'bar'>;
```

If yes, I don't think it's enough to warrant a new type, since the construction above is simple. If it would be useful, we could maybe add the opposite of `MarkRequired` -- `MarkOptional`, which would make the given keys optional.

**Comment by DomParfitt** [comment]
Yeah, that works in the same manner and I think it's also more readable than my own implementation. Thanks!

If you don't think it warrants a new type then I'm fine with that but I'm happy to do the leg work to add it if it wants adding. 🙂

**Comment by krzkaczor** [comment]
@DomParfitt Thanks for this idea. Usually what we do is we keep such issues open for some time to see if there any other people interested in this.

**Comment by krzkaczor** [comment]
@all-contributors please add @DomParfitt for ideas

**Comment by kalbert312** [comment]
I also had a use case for the `MarkOptional` type when supplying default values for an interface that requires them.

I'm using:
```typescript
type MakeOptional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
```

**Comment by krzkaczor** [comment]
Done by @macbem, it's called `MarkOptional` 👌 

Released as `ts-essentials@3.0.5`.
